source/idf/EthernetDevice.cpp

```diff
@@ -0,0 +1,120 @@
+#include "idf/EthernetDevice.hh"
+#include "idf/IOException.hh"
+
+#include <errno.h>
+#include <cstring>
+#include <iostream>
+#include <sstream>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+namespace idf {
+
+EthernetDevice::EthernetDevice(const std::string& id, unsigned length) :
+    InputDevice(id),
+    packetLength(length),
+    sockType(SOCK_STREAM),
+    tcp(true) {
+        memset(&serverAddr, 0 , sizeof(serverAddr));
+        
+    }
+
+void EthernetDevice::open() {
+    if (!mOpen) {
+
+        std::ostringstream stream;
+        stream << "[IDF::EthernetDevice::open()] ";
+
+        socketHandle = socket(AF_INET, sockType, 0);
+
+        if ( errno > 0) {
+            stream << "failed to create socket";
+            perror(stream.str().c_str());
+            throw IOException(stream.str());
+        }
+
+        serverAddr.sin_family = AF_INET;
+        serverAddr.sin_port = htons(serverPort);
+        serverAddr.sin_addr.s_addr = inet_addr(serverName.c_str());
+        serverAddrLen = sizeof(serverAddr);
+
+        std::cout << "[IDF::EthernetDevice] Connecting to " << serverName << ":" << serverPort << std::endl;
+
+        if(tcp) {
+            if (connect(socketHandle, (struct sockaddr*)&serverAddr, serverAddrLen) < 0) {
+                stream << "failed to connect to TCP device " << serverName << ":" << serverPort;
+                perror(stream.str().c_str());
+                throw IOException(stream.str());
+            }
+        } else {
+            char sendBuffer[] = "hello HC";
+            if(sendto(socketHandle, sendBuffer, sizeof(sendBuffer), 0, (struct sockaddr *)&serverAddr, serverAddrLen) < 0) {
```

Comment:
even though this is blocking, the call can still return non-zero due to interruption. Need to handle the case where `errno == EINTR`

Reply:
creativity points for writing the same comment a different way each time.

Reply:
done